File: tools/agentstream/control/control.go
// Package control parses explicit agentstream checklist control trailers.
// REQ: REQ-GOAGENT-CHECKLIST-CONTROL
// ARCH: ARCH-GOAGENT-CHECKLIST-CONTROL
// IMPL: IMPL-GOAGENT-CHECKLIST-CONTROL
package control

import (
	"encoding/json"
	"fmt"
	"strings"
)

const (
	SchemaVersion = 1
	ActionGoto    = "goto"
)

// Decision is the machine-readable routing payload emitted by checklist turns.
type Decision struct {
	SchemaVersion int      `json:"schema_version"`
	Action        string   `json:"action"`
	Target        string   `json:"target,omitempty"`
	Reason        string   `json:"reason,omitempty"`
	Evidence      []string `json:"evidence,omitempty"`
}

type envelope struct {
	Decision Decision `json:"agentstream_control"`
}

// Parse scans fenced JSON blocks from latest to earliest and returns the first
// strict agentstream_control envelope. Prose outside fenced JSON is ignored.
func Parse(text string) (Decision, bool, error) {
	blocks := fencedJSONBlocks(text)
	for i := len(blocks) - 1; i >= 0; i-- {
		var raw map[string]json.RawMessage
		if err := json.Unmarshal([]byte(blocks[i]), &raw); err != nil {
			if strings.Contains(blocks[i], "agentstream_control") {
				return Decision{}, false, fmt.Errorf("invalid agentstream_control JSON: %w", err)
			}
			continue
		}
		body, ok := raw["agentstream_control"]
		if !ok {
			continue
		}
		var env envelope
		env.Decision = Decision{}
		if err := json.Unmarshal(body, &env.Decision); err != nil {
			return Decision{}, false, fmt.Errorf("invalid agentstream_control payload: %w", err)
		}
		return env.Decision, true, nil
	}
	return Decision{}, false, nil
}

// Validate checks a parsed decision against the loaded checklist slug set.
func Validate(d Decision, knownSlugs map[string]bool) error {
	if d.SchemaVersion != SchemaVersion {
		return fmt.Errorf("agentstream_control schema_version must be %d, got %d", SchemaVersion, d.SchemaVersion)
	}
	switch d.Action {
	case ActionGoto:
		if strings.TrimSpace(d.Target) == "" {
			return fmt.Errorf("agentstream_control goto requires target")
		}
		if !knownSlugs[d.Target] {
			return fmt.Errorf("agentstream_control target not found in checklist turns: %s", d.Target)
		}
	default:
		return fmt.Errorf("unsupported agentstream_control action: %s", d.Action)
	}
	return nil
}

func fencedJSONBlocks(text string) []string {
	lines := strings.Split(text, "\n")
	var blocks []string
	inFence := false
	var current []string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if !inFence {
			if trimmed == "```json" {
				inFence = true
				current = current[:0]
			}
			continue
		}
		if strings.HasPrefix(trimmed, "```") {
			blocks = append(blocks, strings.Join(current, "\n"))
			inFence = false
			current = nil
			continue
		}
		current = append(current, line)
	}
	return blocks
}

File: mcp-server/packages/agentstream/src/control.ts
/**
 * [IMPL-GOAGENT-CHECKLIST-CONTROL] [ARCH-GOAGENT-CHECKLIST-CONTROL] [REQ-GOAGENT-CHECKLIST-CONTROL]
 * Parse agentstream_control trailers from agent output (Go control parity, Phase 4a live run).
 */

export const CONTROL_SCHEMA_VERSION = 1;
export const ACTION_GOTO = "goto";

export type ControlDecision = {
  schemaVersion: number;
  action: string;
  target?: string;
  reason?: string;
  evidence?: string[];
};

function fencedJsonBlocks(text: string): string[] {
  const lines = text.split("\n");
  const blocks: string[] = [];
  let inFence = false;
  let current: string[] = [];
  for (const line of lines) {
    const trimmed = line.trim();
    if (!inFence) {
      if (trimmed === "```json") {
        inFence = true;
        current = [];
      }
      continue;
    }
    if (trimmed.startsWith("```")) {
      blocks.push(current.join("\n"));
      inFence = false;
      current = [];
      continue;
    }
    current.push(line);
  }
  return blocks;
}

export function parseControl(
  text: string,
): { decision: ControlDecision; ok: boolean; error?: Error } {
  const blocks = fencedJsonBlocks(text);
  for (let i = blocks.length - 1; i >= 0; i--) {
    const block = blocks[i]!;
    let raw: Record<string, unknown>;
    try {
      raw = JSON.parse(block) as Record<string, unknown>;
    } catch (err) {
      if (block.includes("agentstream_control")) {
        return {
          decision: emptyDecision(),
          ok: false,
          error: new Error(`invalid agentstream_control JSON: ${String(err)}`),
        };
      }
      continue;
    }
    if (!Object.prototype.hasOwnProperty.call(raw, "agentstream_control")) {
      continue;
    }
    const body = raw.agentstream_control as Record<string, unknown>;
    const decision: ControlDecision = {
      schemaVersion: Number(body.schema_version ?? 0),
      action: String(body.action ?? ""),
      target: body.target !== undefined ? String(body.target) : undefined,
      reason: body.reason !== undefined ? String(body.reason) : undefined,
      evidence: Array.isArray(body.evidence)
        ? body.evidence.map((e) => String(e))
        : undefined,
    };
    return { decision, ok: true };
  }
  return { decision: emptyDecision(), ok: false };
}

function emptyDecision(): ControlDecision {
  return { schemaVersion: 0, action: "" };
}

export function validateControl(
  d: ControlDecision,
  knownSlugs: Record<string, boolean>,
): Error | null {
  if (d.schemaVersion !== CONTROL_SCHEMA_VERSION) {
    return new Error(
      `agentstream_control schema_version must be ${CONTROL_SCHEMA_VERSION}, got ${d.schemaVersion}`,
    );
  }
  switch (d.action) {
    case ACTION_GOTO: {
      const target = (d.target ?? "").trim();
      if (target === "") {
        return new Error("agentstream_control goto requires target");
      }
      if (!knownSlugs[target]) {
        return new Error(
          `agentstream_control target not found in checklist turns: ${target}`,
        );
      }
      return null;
    }
    default:
      return new Error(`unsupported agentstream_control action: ${d.action}`);
  }
}
